File: 2.그리디/Programmers/(2019 KAKAO BLIND RECRUITMENT)무지의 먹방 라이브/[jeunghun] solution.py
import heapq

def solution(food_times, k):
    if sum(food_times) <= k:
        return -1
    
    length = len(food_times)
    arr = []    # 우선순위 큐를 저장하기 한한 리스트
    for i in range(n):
        heapq.heappush(arr, (food_times[i], i + 1))
    
    previous = 0    # 이전 음식을 먹는데 걸린 시간
    sum_times = 0    # 총 걸린 시간
    while sum_times + (arr[0][0] - previous) * length <= k:
        target = heapq.heappop(arr)[0]    # 가장 시간이 적게 걸리는 것부터 처리
        sum_times += (target - previous) * length
        length -= 1
        previous = target

    arr = sorted(arr, key=lambda x: x[1])
    return arr[(k - sum_times) % length][1]

File: 2.그리디/Programmers/(2019 KAKAO BLIND RECRUITMENT)무지의 먹방 라이브/[jeunghun] solution.swift
import Foundation

func solution(_ food_times: [Int], _ k: Int64) -> Int {
    let total = food_times.reduce(0) { $0 + $1 }
    
    if total <= k { return -1 }
    
    var length = food_times.count
    var arr: [(Int, Int)] = []
    for i in 0..<length {
        arr.append((food_times[i], i + 1))
    }
    
    var sortedArr = arr.sorted { $0.0 < $1.0 }
    var i = 0
    var previous = 0
    var sumTimes = 0
    
    while sumTimes + (sortedArr[i].0 - previous) * length <= Int(k) {
        let now = sortedArr[i].0
        sumTimes += (now - previous) * length
        previous = now
        length -= 1
        i += 1
    }
    
    sortedArr = sortedArr[i...].sorted { $0.1 < $1.1 }
    return sortedArr[(Int(k) - sumTimes) % length].1
}